File: regions/system/library/dart/config/lib/src/merge.dart
import 'package:yaml/yaml.dart';

/// YamlMap / YamlList を通常の Map / List に変換する。
/// leaf 値はそのまま返す。
dynamic yamlToMap(dynamic yaml) {
  if (yaml is YamlMap) {
    return yaml.map((k, v) => MapEntry(k.toString(), yamlToMap(v)));
  }
  if (yaml is YamlList) {
    return yaml.map(yamlToMap).toList();
  }
  return yaml;
}

/// base マップに overlay マップを再帰的にマージする。
Map<String, dynamic> deepMerge(
    Map<String, dynamic> base, Map<String, dynamic> overlay) {
  final result = Map<String, dynamic>.from(base);
  for (final key in overlay.keys) {
    if (result.containsKey(key) &&
        result[key] is Map &&
        overlay[key] is Map) {
      result[key] = deepMerge(
        result[key] as Map<String, dynamic>,
        overlay[key] as Map<String, dynamic>,
      );
    } else {
      result[key] = overlay[key];
    }
  }
  return result;
}

File: regions/system/library/go/config/merge.go
package config

import (
	"fmt"
	"os"

	"gopkg.in/yaml.v3"
)

// mergeFromFile は envPath の YAML を読み込み、cfg にマージする。
func mergeFromFile(cfg *Config, envPath string) error {
	envData, err := os.ReadFile(envPath)
	if err != nil {
		return fmt.Errorf("failed to read env config: %w", err)
	}

	// base を YAML Value として再パースする
	baseData, err := yaml.Marshal(cfg)
	if err != nil {
		return fmt.Errorf("failed to marshal base config: %w", err)
	}

	var baseValue map[string]interface{}
	if err := yaml.Unmarshal(baseData, &baseValue); err != nil {
		return fmt.Errorf("failed to parse base config: %w", err)
	}

	var envValue map[string]interface{}
	if err := yaml.Unmarshal(envData, &envValue); err != nil {
		return fmt.Errorf("failed to parse env config: %w", err)
	}

	merged := deepMerge(baseValue, envValue)

	mergedData, err := yaml.Marshal(merged)
	if err != nil {
		return fmt.Errorf("failed to marshal merged config: %w", err)
	}

	if err := yaml.Unmarshal(mergedData, cfg); err != nil {
		return fmt.Errorf("failed to parse merged config: %w", err)
	}

	return nil
}

// deepMerge は base マップに overlay マップを再帰的にマージする。
func deepMerge(base, overlay map[string]interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	for k, v := range base {
		result[k] = v
	}
	for k, v := range overlay {
		if baseVal, ok := result[k]; ok {
			baseMap, baseIsMap := baseVal.(map[string]interface{})
			overlayMap, overlayIsMap := v.(map[string]interface{})
			if baseIsMap && overlayIsMap {
				result[k] = deepMerge(baseMap, overlayMap)
				continue
			}
		}
		result[k] = v
	}
	return result
}
